[cmake] Fixes / IOS / Jenkins

fetzerch opened the pull request:
This contains the CMake work of the previous week. Even though the topics are unrelated, the commits are dependent, so for simplicity I'd prefer to push it together (unless you want me to split).
- First 3 commits are just build fixes (ping @wsnipex to avoid that we do duplicate work)
- Most of the commits are related to building kodi.bin on IOS. Note that this is untested as packaging is still missing.
- Last commits add a CMAKE_BUILD option to the buildscripts used by Jenkins. We can now build kodi.bin/exe on Jenkins for Linux, Rbpi, OSX, IOS, Android Arm and Android X86 (That might be interesting for @MartijnKaijser)

@Memphiz:
- Could you add a CMAKE_BUILD parameter to Jenkins so that I can drop the last commit.
- You might also be interested in reviewing 7de0e0b
- Do you have any idea how to fix the IOS build on jenkins (http://jenkins.kodi.tv/job/IOS/6755/consoleFull)? Maybe @wsnipex knows as it's dvdnav related. Unfortunately I can't reproduce that error locally so I'm stuck here.

Anyone with windows batch scripting skills interested in adding CMake support to the windows build script used by Jenkins? Maybe @afedchin or @AchimTuran? :)

fetzerch committed:
[cmake] Modernize FindOpenGl.cmake

fetzerch committed:
[cmake] Modernize FindVAAPI.cmake

fetzerch committed:
[cmake] Modernize FindVDPAU.cmake

fetzerch committed:
[jenkins] Optionally build with CMake

- Add CMAKE_BUILD variable that defines if CMake should be used
  to build the main application.
- Use variable to build with CMake (GNU Makefile Generator) on
  supported platforms: android, ios, osx, linux, rbpi.

fetzerch committed:
[cmake] Sync DVDVideoCodecVDA.h/cpp removal

fetzerch committed:
[cmake] Build kodi.bin for ios

First step for ios support with CMake.

fetzerch committed:
[jenkins] Temporarily disable packaging for CMake

Packaging is currently not supported when building with CMake.

fetzerch committed:
[cmake] Add missing CPU/ARCH combinations to archsetup scripts

fetzerch committed:
[cmake] LibDvdCSS is an optional dependency, LibDvd is required

fetzerch committed:
[toolchain] Ensure CMAKE_CROSSCOMPILING is set for IOS

Setting CMAKE_SYSEM_NAME in a Toolchain file tells CMake that we're cross
compiling. This is needed for IOS to make sure that the native tools are not
built when building Kodi with CMake.

fetzerch committed:
[cmake] Add option for VTB decoding

fetzerch committed:
[cmake] Define DHAVE_LIBEGL=1 in FindOpenGLES.cmake module

The FindOpenGLES module searches for EGL hence should also define HAVE_LIBEGL.

fetzerch committed:
[ios] Prefix includes and imports with the folder

This way we don't have to add the whole xbmc tree to the include directories.

fetzerch committed:
[cmake] Use variable for main.cpp

The different platforms can use different main.cpp files. As currently still
all platforms except IOS use the posix main.cpp, the variable defaults to that
but can be overridden in the platform specific archsetup.cmake files.

fetzerch committed:
[cmake] Add CMake files for library.kodi.audioengine

fetzerch committed:
[cmake] Sync InputStream project files

jjd-uk commented:
@Paxxi maybe able to help with Windows build script.

Memphiz commented:
@fetzerch added CMAKE_BUILD to all jenkins jobs defaulting to false. Also the include fixes look good and are very welcome :)

wsnipex commented on tools/depends/target/cmakebuildsys/Makefile:
```diff
@@ -0,0 +1,16 @@
+
+all: $(XBMCROOT)/lib$(APP_NAME).so
+
+$(XBMCROOT)/lib$(APP_NAME).so:
```
this will only work for platforms that build libkodi.so or gtests.

fetzerch replied:
Ah true, do you have a suggestion how it should be done? Depending on what CMake generator we use it might generate different output. (with GNU Makefiles we build ./kodi.bin but with Xcode it builds it in a Debug/Release subfolder. Could we just make it a phony target?

wsnipex replied:
phony would be ok for now, since cmake will take to not needlessly rebuild anyway.

wsnipex commented on xbmc/cores/AudioEngine/CMakeLists.txt:
```diff
@@ -95,27 +95,28 @@ if(CORE_SYSTEM_NAME STREQUAL rbpi)
 
-if(CORE_SYSTEM_NAME STREQUAL darwin OR CORE_SYSTEM_NAME STREQUAL ios)
-  list(APPEND SOURCES Sinks/osx/AEDeviceEnumerationOSX.cpp
+if(CORE_SYSTEM_NAME STREQUAL darwin)
```
CMAKE_SYSTEM_NAME

fetzerch replied:
`CMAKE_SYSTEM_NAME` is Darwin for all Apple platforms (osx, ios, tvos).
`CORE_SYSTEM_NAME` is darwin for osx but it's set to ios for ios. Memphiz already told me that we should change `CORE_SYSTEM_NAME` to osx at some point to be consistent. Here we really mean `if(osx)` hence it should be correct the way it is.

wsnipex replied:
ah ok, I though it was already changed so CORE_SYSTEM_NAME=osx
sorry for the noise.

wsnipex commented:
very nice!

fetzerch committed:
[cmake] Allow building tests when cross compiling

Tests can be built when cross compiling (but they can't be executed
with CTest).

fetzerch committed:
[cmake] Fix linking libdvdnav

Remove the following warnings and fixes IOS build:
- linking shared library against static library is not portable
- file was built for archive which is not the architecture being linked (x86_64)

fetzerch committed:
[cmake] Add cross platform whole_archive function

Add a function that wraps list of libraries with --whole_archive or -force_read
depending on the compiler toolchain.

This fixes the unit tests on OSX as otherwise an empty test binary is created.

fetzerch commented:
Thanks for reviewing. I fixed the build error on IOS (libdvdnav needs to link against libdvdread.la instead .a) and the tests are working as well now on OSX-64 and Linux-64: http://jenkins.kodi.tv/view/Helpers/job/BuildMulti-All/1241/
